rhel/vex/fetcher.go

```diff
@@ -0,0 +1,390 @@
+	success = true
+	return f, driver.Fingerprint(fp.String()), nil
+}
+
+// ProcessChanges deals with the published changes.csv, adding records
+// to w means they are deemed to have changed since the compressed
+// file was last processed. w and fp can be modified.
+func (u *Updater) processChanges(ctx context.Context, w io.Writer, fp *fingerprint) error {
+	uri, err := u.url.Parse(changesFile)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, uri.String(), nil)
+	if err != nil {
+		return err
+	}
+	if fp.changesEtag != "" {
+		req.Header.Add("If-None-Match", fp.changesEtag)
+	}
+	res, err := u.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	switch res.StatusCode {
+	case http.StatusOK:
+		if t := fp.changesEtag; t == "" || t != res.Header.Get("etag") {
+			break
+		}
+		fallthrough
+	case http.StatusNotModified:
+		return nil
+	default:
+		return fmt.Errorf("unexpected response from changes.csv: %s", res.Status)
+	}
+	fp.changesEtag = res.Header.Get("etag")
+
+	rd := csv.NewReader(res.Body)
+	rd.FieldsPerRecord = 2
+	rd.ReuseRecord = true
+	var (
+		l       int
+		buf, bc bytes.Buffer
+	)
+	rec, err := rd.Read()
+	for ; err == nil; rec, err = rd.Read() {
+		buf.Reset()
+		bc.Reset()
+		if len(rec) != 2 {
+			return fmt.Errorf("could not parse changes.csv file")
+		}
+
+		cvePath, uTime := rec[0], rec[1]
+		year, err := strconv.ParseInt(path.Dir(cvePath), 10, 64)
+		if err != nil {
+			return fmt.Errorf("error parsing year %w", err)
+		}
+		if year < lookBackToYear {
+			continue
+		}
+		updatedTime, err := time.Parse(time.RFC3339, uTime)
+		if err != nil {
+			return fmt.Errorf("line %d: %w", l, err)
+		}
+		if updatedTime.Before(fp.requestTime) {
+			continue
+		}
+
+		advisoryURI, err := u.url.Parse(cvePath)
+		if err != nil {
+			return err
+		}
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, advisoryURI.String(), nil)
+		if err != nil {
+			return fmt.Errorf("error creating advisory request %w", err)
+		}
+
+		// Use a func here as we're in a loop and want to make sure the
+		// body is closed in all events.
+		err = func() error {
+			res, err := u.client.Do(req)
+			if err != nil {
+				return fmt.Errorf("error making advisory request %w", err)
+			}
+			defer res.Body.Close()
+			err = httputil.CheckResponse(res, http.StatusOK)
+			if err != nil {
+				return fmt.Errorf("unexpected response: %w", err)
+			}
+
+			// Add compacted JSON to buffer.
+			_, err = buf.ReadFrom(res.Body)
+			if err != nil {
+				return fmt.Errorf("error reading from buffer: %w", err)
+			}
+			zlog.Debug(ctx).Str("url", advisoryURI.String()).Msg("copying body to file")
+			err = json.Compact(&bc, buf.Bytes())
+			if err != nil {
+				return fmt.Errorf("error compressing JSON: %w", err)
+			}
+
+			bc.WriteByte('\n')
+			w.Write(bc.Bytes())
+			l++
+			return nil
+		}()
+		if !errors.Is(err, nil) {
+			return err
+		}
+
+	}
+
+	switch {
+	case errors.Is(err, nil):
```

Comment:
In this case it wouldn't but in the changes function it would as the `csv.Read()` err is only scoped to the for loop in that function. My preference is to revert the change that moves the initial `Read()` to the for loop declaration and then just check if it's EOF or not.